python-package/lightgbm/basic.py

```diff
@@ -1174,6 +1175,25 @@ def num_feature(self):
         else:
             raise LightGBMError("Cannot get num_feature before construct dataset")
 
+    def get_ref_chain(self, ref_limit = 100):
+        '''
+        Gets a chain of Dataset objects, starting with r, then going to r.reference if exists, 
+            then to r.reference.reference, etc. until we hit ref_limit or a reference loop
+
+        Returns
+        -------
+        chain of references of self : set of Dataset objects
+        '''
+        head = self
+        ref_chain=[]
+        while len(ref_chain)<ref_limit:
+            ref_chain += ([head] if isinstance(head, Dataset) else [])
+            if isinstance(head, Dataset) and (head.reference is not None) and (head.reference not in ref_chain):
```

Comment:
So I made this change and re-built but excluding things with handle = None breaks the example in #714.  In particular, all of tmp_dat, tmp_dat_train, tmp_dat_val have handle=None.  To be totally honest I'm not super clear on what handle is.  Is there an easy way around this?